File: collector/internal/validator/span.go
package validator

import (
	"fmt"
	"time"
)

type SpanEvent struct {
	ID            string            `json:"id"`
	TraceID       string            `json:"trace_id"`
	ParentID      string            `json:"parent_id,omitempty"`
	ServiceName   string            `json:"service_name"`
	OperationName string            `json:"operation_name"`
	Kind          string            `json:"kind"`
	StartedAt     time.Time         `json:"started_at"`
	EndedAt       *time.Time        `json:"ended_at,omitempty"`
	DurationMs    *int64            `json:"duration_ms,omitempty"`
	Status        string            `json:"status"`
	Error         *SpanError        `json:"error,omitempty"`
	Tags          map[string]string `json:"tags,omitempty"`
	Logs          []EmbeddedLog     `json:"logs,omitempty"`
	WorkspaceID   string            `json:"workspace_id"`
}

type SpanError struct {
	Type    string `json:"type"`
	Message string `json:"message"`
	Code    string `json:"code,omitempty"`
}

var validKinds = map[string]bool{
	"server":   true,
	"client":   true,
	"producer": true,
	"consumer": true,
	"internal": true,
}

var validStatuses = map[string]bool{
	"ok":          true,
	"error":       true,
	"timeout":     true,
	"in_progress": true,
}

func Validate(s *SpanEvent) error {
	if s.ID == "" {
		return fmt.Errorf("id is required")
	}
	if s.TraceID == "" {
		return fmt.Errorf("trace_id is required")
	}
	if s.ServiceName == "" {
		return fmt.Errorf("service_name is required")
	}
	if s.OperationName == "" {
		return fmt.Errorf("operation_name is required")
	}
	if !validKinds[s.Kind] {
		return fmt.Errorf("kind must be one of: server, client, producer, consumer, internal")
	}
	if s.StartedAt.IsZero() {
		return fmt.Errorf("started_at is required")
	}
	if !validStatuses[s.Status] {
		return fmt.Errorf("status must be one of: ok, error, timeout, in_progress")
	}
	if s.WorkspaceID == "" {
		return fmt.Errorf("workspace_id is required")
	}
	return nil
}

type LogEvent struct {
	ID          string            `json:"id"`
	TraceID     string            `json:"trace_id"`
	ServiceName string            `json:"service_name"`
	Level       string            `json:"level"`
	Message     string            `json:"message"`
	Attributes  map[string]string `json:"attributes,omitempty"`
	Timestamp   time.Time         `json:"timestamp"`
	WorkspaceID string            `json:"workspace_id"`
}

var validLogLevels = map[string]bool{
	"DEBUG": true,
	"INFO":  true,
	"WARN":  true,
	"ERROR": true,
}

func ValidateLog(l *LogEvent) error {
	if l.TraceID == "" {
		return fmt.Errorf("trace_id is required")
	}
	if l.ServiceName == "" {
		return fmt.Errorf("service_name is required")
	}
	if !validLogLevels[l.Level] {
		return fmt.Errorf("level must be one of: DEBUG, INFO, WARN, ERROR")
	}
	if l.Message == "" {
		return fmt.Errorf("message is required")
	}
	if l.WorkspaceID == "" {
		return fmt.Errorf("workspace_id is required")
	}
	return nil
}

type EmbeddedLog struct {
	Level      string            `json:"level"`
	Message    string            `json:"message"`
	Attributes map[string]string `json:"attributes,omitempty"`
	Timestamp  string            `json:"timestamp"`
}

File: sdk-node/src/span.ts
import { SpanEvent, SpanError, SpanKind, SpanStatus, SpanLog } from './types'
import { generateSpanId } from './id'

export class Span {
  readonly id: string
  readonly traceId: string
  readonly parentId?: string
  readonly serviceName: string
  readonly operationName: string
  readonly kind: SpanKind
  readonly workspaceId?: string
  readonly startedAt: Date

  private endedAt?: Date
  private _status: SpanStatus = 'in_progress'
  private _error?: SpanError
  private _tags: Record<string, string> = {}

  private _onEnd?: (span: Span) => void

  private _logs: SpanLog[] = []

  constructor(opts: {
    traceId: string
    parentId?: string
    serviceName: string
    operationName: string
    kind: SpanKind
    workspaceId?: string
    tags?: Record<string, string>
    onEnd?: (span: Span) => void
  }) {
    this.id = generateSpanId()
    this.traceId = opts.traceId
    this.parentId = opts.parentId
    this.serviceName = opts.serviceName
    this.operationName = opts.operationName
    this.kind = opts.kind
    this.workspaceId = opts.workspaceId
    this._tags = opts.tags ?? {}
    this._onEnd = opts.onEnd
    this.startedAt = new Date()
  }

  setTag(key: string, value: string): this {
    this._tags[key] = value
    return this
  }

  setError(err: Error | SpanError): this {
    this._status = 'error'
    if (err instanceof Error) {
      this._error = { type: err.name, message: err.message }
    } else {
      this._error = err
    }
    return this
  }

  setTimeout(): this {
    this._status = 'timeout'
    return this
  }

  log(message: string, attributes?: Record<string, unknown>, level: 'DEBUG' | 'INFO' | 'WARN' | 'ERROR' = 'INFO'): this {
    const flat: Record<string, string> = {}
    if (attributes) {
      for (const [k, v] of Object.entries(attributes)) {
        if (v === null || v === undefined) continue
        flat[k] = typeof v === 'object' ? JSON.stringify(v) : String(v)
      }
    }
    this._logs.push({ level, message, attributes: flat, timestamp: new Date().toISOString() })
    return this
  }

  end(status?: SpanStatus): void {
    this.endedAt = new Date()
    if (status) this._status = status
    else if (this._status === 'in_progress') this._status = 'ok'
    this._onEnd?.(this)
  }

  toEvent(): SpanEvent {
    const durationMs = this.endedAt
      ? this.endedAt.getTime() - this.startedAt.getTime()
      : undefined

    return {
      id: this.id,
      trace_id: this.traceId,
      parent_id: this.parentId,
      service_name: this.serviceName,
      operation_name: this.operationName,
      kind: this.kind,
      started_at: this.startedAt.toISOString(),
      ended_at: this.endedAt?.toISOString(),
      duration_ms: durationMs,
      status: this._status,
      error: this._error,
      tags: this._tags,
      logs: this._logs,
      ...(this.workspaceId ? { workspace_id: this.workspaceId } : {}),
    }
  }
}
